### ISSUE TITLE
Alluxio file locality reported as all workers

### ISSUE THREAD

**Issue by tgravescs** [NVIDIA/spark-rapids#6530]
Alluxio is returning hostnames for all the workers when looking at the block locations returned.  Note that 
alluxio.user.ufs.block.location.all.fallback.enabled=true by default, setting that to false make it return nothing.  But I was expecting alluxio:// path to return the hostname where the blocks are actually cached. That doesn't seem to be happening on our Databricks setup so not sure if its supposed to e and we have something configured wrong or if it just never returns the block location of the cached block.  Need to investigate.

**Comment by res-life** [comment]
I tested, and the behavior is what we expect.
By default `alluxio.user.ufs.block.location.all.fallback.enabled` is true.
## Scenario 1: the property is true:
  If there are no workers that have the requested cache, will return all workers.
  If there are any workers that have the requested cache, will return the workers where the blocks are actually cached.
## Scenario 2: the property is false:
   Return the workers where the blocks are actually cached.

My test code is:
```
    InstancedConfiguration conf = InstancedConfiguration.defaults();
    conf.set(PropertyKey.USER_UFS_BLOCK_LOCATION_ALL_FALLBACK_ENABLED, "true"); // or false
    FileSystem fs = FileSystem.Factory.create(conf);
    String url = "alluxio://xxx.parquet";
    AlluxioURI path = new AlluxioURI(url);
    List<BlockLocationInfo> locations = fs.getBlockLocations(path);
    System.out.println(locations);
```

The property is client-side property.
The client-side code will add all worker nodes if it's true and the locations are empty.

See the [code]:(https://github.com/Alluxio/alluxio/blob/v2.8.0/core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java#L239ToL243), 
```
if (locations.isEmpty() && mFsContext.getPathConf(path)
    .getBoolean(PropertyKey.USER_UFS_BLOCK_LOCATION_ALL_FALLBACK_ENABLED)) {
  // Case 2: Fallback to add all workers to locations so some apps (Impala) won't panic.
  locations.addAll(getHostWorkerMap().values());
  Collections.shuffle(locations);
}
```

Did you require an un-cached block? If yes, it's right to return all the workers when the property is true. And it's right to return empty when the property is false.

For the output of `alluxio fs stat file-path`, we can see some blocks are cached, and some are not.

```
$ ./bin/alluxio fs stat  /ndsv2-data/parquet_sf3000/web_sales/ws_sold_date_sk=2452047/part-00162-d67a24e4-5fd1-4729-9c27-a54516dea8f1.c000.snappy.parquet
/ndsv2-data/parquet_sf3000/web_sales/ws_sold_date_sk=2452047/part-00162-d67a24e4-5fd1-4729-9c27-a54516dea8f1.c000.snappy.parquet is a file path.
URIStatus{info=FileInfo{fileId=246289530879, name=part-00162-d67a24e4-5fd1-4729-9c27-a54516dea8f1.c000.snappy.parquet, path=/ndsv2-data/parquet_sf3000/web_sales/ws_sold_date_sk=2452047/part-00162-d67a24e4-5fd1-4729-9c27-a54516dea8f1.c000.snappy.parquet, ufsPath=s3a://ndsv2-data/parquet_sf3000/web_sales/ws_sold_date_sk=2452047/part-00162-d67a24e4-5fd1-4729-9c27-a54516dea8f1.c000.snappy.parquet, length=50335374, blockSizeBytes=16777216, creationTimeMs=1663135370561, completed=true, folder=false, pinned=false, pinnedlocation=[], cacheable=true, persisted=true, blockIds=[246272753664, 246272753665, 246272753666, 246272753667], inMemoryPercentage=0, lastModificationTimesMs=1658241959000, ttl=-1, lastAccessTimesMs=1663135551850, ttlAction=DELETE, owner=ubuntu, group=ubuntu, mode=493, persistenceState=PERSISTED, mountPoint=false, replicationMax=1, replicationMin=0, fileBlockInfos=[FileBlockInfo{blockInfo=BlockInfo{id=246272753664, length=16777216, locations=[]}, offset=0, ufsLocations=[]}, FileBlockInfo{blockInfo=BlockInfo{id=246272753665, length=16777216, locations=[]}, offset=16777216, ufsLocations=[]}, FileBlockInfo{blockInfo=BlockInfo{id=246272753666, length=16777216, locations=[BlockLocation{workerId=3093864047445409794, address=WorkerNetAddress{host=10.59.242.224, containerHost=, rpcPort=29999, dataPort=29999, webPort=30000, domainSocketPath=/local_disk0/alluxio_domain_socket/0eb0410d-dab4-46c7-bbff-fde1da7d0a56, tieredIdentity=TieredIdentity(node=10.59.242.224, rack=null)}, tierAlias=SSD, mediumType=SSD}]}, offset=33554432, ufsLocations=[]}, FileBlockInfo{blockInfo=BlockInfo{id=246272753667, length=3726, locations=[BlockLocation{workerId=3093864047445409794, address=WorkerNetAddress{host=10.59.242.224, containerHost=, rpcPort=29999, dataPort=29999, webPort=30000, domainSocketPath=/local_disk0/alluxio_domain_socket/0eb0410d-dab4-46c7-bbff-fde1da7d0a56, tieredIdentity=TieredIdentity(node=10.59.242.224, rack=null)}, tierAlias=SSD, mediumType=SSD}]}, offset=50331648, ufsLocations=[]}], mountId=5240861571657171002, inAlluxioPercentage=33, ufsFingerprint=TYPE|FILE UFS|s3 OWNER|_ GROUP|_ MODE|493 CONTENT_HASH|37c32c51592baf02b1f789d4e171c48f , acl=user::rwx,group::r-x,other::r-x, defaultAcl=}, cacheContext=null}
Containing the following blocks: 
BlockInfo{id=246272753664, length=16777216, locations=[]}
BlockInfo{id=246272753665, length=16777216, locations=[]}
BlockInfo{id=246272753666, length=16777216, locations=[BlockLocation{workerId=3093864047445409794, address=WorkerNetAddress{host=10.59.242.224, containerHost=, rpcPort=29999, dataPort=29999, webPort=30000, domainSocketPath=/local_disk0/alluxio_domain_socket/0eb0410d-dab4-46c7-bbff-fde1da7d0a56, tieredIdentity=TieredIdentity(node=10.59.242.224, rack=null)}, tierAlias=SSD, mediumType=SSD}]}
BlockInfo{id=246272753667, length=3726, locations=[BlockLocation{workerId=3093864047445409794, address=WorkerNetAddress{host=10.59.242.224, containerHost=, rpcPort=29999, dataPort=29999, webPort=30000, domainSocketPath=/local_disk0/alluxio_domain_socket/0eb0410d-dab4-46c7-bbff-fde1da7d0a56, tieredIdentity=TieredIdentity(node=10.59.242.224, rack=null)}, tierAlias=SSD, mediumType=SSD}]}
```

Note: Usually the first block is not cached and the tail blocks are cached if cache capacity is relatively small because the tail blocks which contain meta are hot blocks.

**Comment by res-life** [comment]
@tgravescs  Help to check

**Comment by tgravescs** [comment]
No, I was specifically testing cached blocks and made sure they were cached via alluxio cli.  The scenarios above are the correct ones. So you are saying your test code properly only returned the locations of only the cached blocks when they were cached?  
```
List<BlockLocationInfo> locations = fs.getBlockLocations(path);
System.out.println(locations);
```
I tried code similar to yours but my main test was just by having our code print the locations as Spark was going to use them.

**Comment by tgravescs** [comment]
where did you run your test code from?

**Comment by tgravescs** [comment]
```
    FileSystem fs = FileSystem.Factory.create(conf);
    String url = "alluxio://xxx.parquet";
    AlluxioURI path = new AlluxioURI(url);
    List<BlockLocationInfo> locations = fs.getBlockLocations(path);
```

What type of filesystem does this give you?
I wonder if its something in our code or the type we get because its originally an s3:// url.  Please try printing the location via our code in like the GpuFileSourceScanExec or GpuParquetScan

**Comment by tgravescs** [comment]
Doing it through a notebook with the hadoop filesystem api does also seem to just the cached worker correctly. I'll try inside our code, otherwise perhaps something else was going on with the cluster.

```
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
val url = new Path("alluxio:/XXX")
val fs = url.getFileSystem(new Configuration())
val res = fs.getFileBlockLocations(url, 0, 10)
```

**Comment by tgravescs** [comment]
ok so with a single file I was not able to reproduce the issue, but with a single file it also put blocks for it on both of my nodes ( using 2 node cluster).  If I read a lot more files, then I was able to reproduce it but I think in this case perhaps the entire file wasn't downloaded (only 24%).  if I ran alluxio fs location on the file it gave me one node but then when I ran the Spark job the block locations came back as both nodes. In this case I was only doing select and showed me the first 1000 rows so it might not have needed the entire file.  I ran a count() on the same dataset and made sure the file was 100% downloaded and then I saw Alluxio report the block locations properly.  So when I reported this I may have been checking files that weren't 100% downloaded yet because I was just using the alluxio fs location command to check files.
